shiny/_autoreload.py

```diff
@@ -231,3 +207,97 @@ async def process_request(
         await asyncio.Future()  # wait forever
+
+
+class ResponseMangler:
+    """A class that assists with intercepting and rewriting response bodies being sent
+    over ASGI. This would be easy if not for 1) response bodies are potentially sent in
+    chunks, over multiple events; 2) the first response event we receive is the one that
+    contains the Content-Length, which can be affected when we do rewriting later on.
+    The ResponseMangler handles the buffering and content-length rewriting, leaving the
+    caller to only have to worry about the actual body-modifying logic.
+    """
+
+    def __init__(
+        self, send: ASGISendCallable, mangler: Callable[[bytes], Tuple[bytes, bool]]
+    ):
+        # The underlying ASGI send function
+        self._send = send
+        # The caller-provided logic for rewriting the body. Takes a single `bytes`
+        # argument that is _all_ of the body bytes seen _so far_, and returns a tuple of
+        # (bytes, bool) where the bytes are the (possibly modified) body bytes and the
+        # bool is True if the mangler does not care to see any more data.
+        self._mangler = mangler
+
+        # If True, the mangler is done and any further data can simply be passed along
+        self._done: bool = False
+
+        # Holds the http.response.start event, which may need its Content-Length header
+        # rewritten before we send it
+        self._response_start: Optional[HTTPResponseStartEvent] = None
+        # All the response body bytes we have seen so far
+        self._body = b""
+
+    async def send(self, event: ASGISendEvent) -> None:
+        if self._done:
+            await self._send(event)
+            return
+
+        if event["type"] == "http.response.start":
+            self._response_start = event
+        elif event["type"] == "http.response.body":
+            # This check is mostly to make pyright happy
+            if self._response_start is None:
+                raise AssertionError(
+                    "http.response.body ASGI event sent before http.response.start"
+                )
+
+            # Add the newly received body data to what we've seen already
+            self._body += event["body"]
+            # Snapshot length before we mess with the body
+            old_len = len(self._body)
+            # Mangle away! If done is True, the mangler doesn't want to do any further
+            # mangling.
+            self._body, done = self._mangler(self._body)
+
+            new_len = len(self._body)
+            if new_len != old_len:
+                # The mangling check changed the length of the body. Add the difference
+                # to the content-length header (if content-length is even present)
+                _add_to_content_length(self._response_start, new_len - old_len)
+
+            more_body = "more_body" in event and event["more_body"]
+
+            if done or not more_body:
```

Comment:
```suggestion
            more_body = event.get("more_body", None)

            if done or more_body is None:
```

Reply:
It's possible for `more_body` to be present but `False`--I'll take the rest of your suggestion though, thanks!